Improve analysis manageable affordances document and add an initial list of user stories. (#2107)
w3c/wot-thing-description: planning/work-items/analysis/analysis-manageable-affordances.md

- **Process Stakeholders**:
  - Submitter: WoT WG/extended community
  - Specification Writers:
  - Implementation Volunteers:
  - Impacted People: TD Designers, Consumer application developers.
  - Impact Type: Increase complexity of the WoT interaction model. This implies an increase in implementation complexity of both Consumer and Exposer applications.
- **Linked Use Cases or Categories:**
  - https://github.com/w3c/wot-thing-description/issues/1070
  - https://github.com/w3c/wot-thing-description/issues/1779

#### Scenarios

- As a customer using a ticket kiosk I want to tell the printer to print three tickets and notify me on the screen when it is complete
- As a factory worker I want to tell a robotic arm to make a series of movements in sequence, without having to wait for each movement to complete before starting the next
- As a factory worker I want to tell a 3D printer to print a series of objects so that I don't have to wait for one object to be completed before submitting the next

> ### Cuncurrent Actions

Reviewer: ```suggestion
### Concurrent Actions
```

- **Who:** Thing Description designer
- **What:** Describe actions that can be executed mutliple times concurrently
- **Why:** So that I can describe devices that can handle multiple running operations at the same time
- **Sentence:** As a Thing Description designer, I need to describe actions that can be executed mutliple times concurrently so that I can describe devices that can handle multiple running operations at the same time.
- **Process Stakeholders**:
  - Submitter: WoT WG/extended community
  - Specification Writers:
  - Implementation Volunteers:
  - Impacted People: TD Designers, Consumer application developers.
  - Impact Type: Increase complexity of the WoT interaction model. This implies an increase in implementation complexity of both Consumer and Exposer applications.

#### Scenarios

- As an office worker using a high end photocopier I want to copy multiple multi-page documents at the same time to save me time

> - As a factory worker I want to instruct a multi-head 3D printer to print two objects at once so that I don't have to wait for one to finish before starting the other

Reviewer: ```suggestion
- As a factory worker I want to instruct a multi-head 3D printer to print two objects at once so that I don't have to wait for one to finish before starting the other
- As two office workers using a high-end automated coffee machine with two brew chambers, we want to brew two coffees at the same time to avoid waiting time
```

### Express dependencies between affordances

- **Who:** Thing Description designer
- **What:** Describe dependencies and relationships between multiple affordances.
- **Why:** So that I can define how affordances are related and constrain how they can be used in relation with one another.
- **Sentence:** As a Thing Description designer, I need to express dependencies between affordances so that I can define valid interaction flows and link related affordances.
- **Process Stakeholders**:
  - Submitter: WoT WG/extended community
  - Specification Writers:
  - Implementation Volunteers:
  - Impacted People: TD Designers, Consumer application developers.

>   - Impact Type: Increase complexity of the WoT interaction model. This implies an increase in implementation complexity of both Consumer and Exposer applications.

Reviewer: ```suggestion
  - Impact Type: Increase complexity of the WoT interaction model. This implies an increase in implementation complexity of both Consumer and Exposer applications. Note: If an existing Thing already has this behavior, there is no increased complexity as the behavior was described out-of-band somehow.
```

## User Stories

### Monitorable and Cancelable Actions

* **Who:** Thing Description designer / Device Manufacturer / System integrator
* **What:** Describe actions that can be monitored and canceled

> * **Why:** I can describe the lifecycle of actions in automation systems

Author: ```suggestion
* **Why:** I can monitor and cancel a long running action without waiting indefinitely for its completion
```

Reviewer 2: "I can monitor" --> "I want to monitor" ???

Reviewer 3: ```suggestion
* **Why:** So that I can describe the lifecycle of actions in automation systems
```

* **Sentence:** As a Thing Description designer, I need to describe actions that can be monitored and canceled so that I can correctly describe the lifecycle of actions in automation systems.
* **Process Stakeholders**:
  - Submitter: Multiple
  - Specification Writers: 
  - Implementation Volunteers: 
  - Impacted People: TD Designers and Consumer application developers.
  - Impact Type: Increase complexity of the WoT interaction model. This implies an increase in implementation complexity of both Consumer and Exposer applications. Finally, the Thing Description design process might result
    more complex too. 
* **Linked Use Cases or Categories:**
    * https://github.com/w3c-cg/webagents/issues/26
    * https://github.com/w3c-cg/webagents/issues/27
    * https://github.com/w3c-cg/webagents/issues/29
    * https://github.com/w3c-cg/webagents/issues/31

### Queue of Actions

* **Who:** Thing Description designer
* **What:** Describe a queue of actions as a resource, including its contents and state

> * **Why:** I can fully describe the current status of a device capable of running persistent actions

Reviewer 2: Is it  "I can inspect" or "I want to inspect" ?

> * **Why:** I can fully describe the current status of a device capable of running persistent actions

Reviewer: the first one. The why is read like "so that I"

> * **Why:** I can fully describe the current status of a device capable of running persistent actions

Reviewer 3: Just say it... Electrons are free, toner and ink are cheap, and these few words (even if repeated in every user story and/or use case) are very unlikely to (substantially) increase printed page counts.
```suggestion
* **Why:** So that I can fully describe the current status of a device capable of running persistent actions
```

* **Why:** So that I can fully describe the current status of a device capable of running persistent actions
* **Sentence:** As a Thing Description designer, I need to describe a queue of actions as a resource so that I can fully describe the current status of a device capable of running monitorable and cancelable actions.
* **Process Stakeholders**:
    * Submitter: WoT WG/extended community
    * Specification Writers:
    * Implementation Volunteers:
    * Impacted People: TD Designers, Consumer application developers.
    * Impact Type: Increase complexity of the WoT interaction model. This implies an increase in implementation complexity of both Consumer and Exposer applications. 
* **Linked Use Cases or Categories:**
   * https://github.com/w3c/wot-thing-description/issues/1070
   * https://github.com/w3c/wot-thing-description/issues/1779

### Express dependencies between affordances

* **Who:** Thing Description designer
* **What:** Describe dependencies and relationships between multiple affordances.

> * **Why:** So that I can define how affordances are related and constrain how they can be used in relation of one another.

Reviewer 3: ```suggestion
* **Why:** So that I can define how affordances are related and constrain how they can be used in relation with one another.
```

- **Support for queue**: Implicitly supported through MQTT topics. But the device might reject execution of multiple commands if it is busy.
- **Reference**: [AWS IoT Commands](https://docs.aws.amazon.com/iot/latest/developerguide/iot-remote-command-concepts.html)

### Particle IoT
- **Support for Asynchronous Actions**: No
- **Description**: Particle IoT does not have a built-in concept for asynchronous actions but only supports synchronous action invocations. Developers can deploy workarounds to achive similar functionality, such as using a combination of events and and functions to simulate asynchronous behavior.
- **Support for queue**: No
- **Reference**: [Particle Functions](https://docs.particle.io/reference/cloud-apis/api/#call-a-function)

> ### Azure IoT Hub 

Reviewer: same comment as for particle iot

## Existing Solutions

### Use `queryaction` and `cancelaction`

To cover the `Monitorable and Cancelable Actions` user story, we can leverage the newly introduced operation in the TD operation set. However, these two operations are currently underspecified, and a Consumer cannot properly understand which payload to send to the `queryaction` operation and what it receives back. Cancellation is also affected by the same problems. In summary, the current solution only works for actions that have a single running instance and do not require complex payload relations between `invokeaction`, `queryaction`, and `cancelaction`. 

### Use a "meta-affordance" to describe extra resources or interaction patterns

> With the current specification, users can support to same level all three user stories by creating ad hoc affordances that mimics the requirements of the underlying interaction pattern. For example, the queue of actions can be represented by a *PropertyAffordance*. [BACnet binding](https://w3c.github.io/wot-binding-templates/bindings/protocols/bacnet/index.html#event-mappings) is also currently expressing relationships between events and actions thanks to a combination of ad hoc affordances and binding defined well-known relationships. 

Reviewer 3: ```suggestion
According to the current specification, users can support all three user stories to the same level by creating ad hoc affordances that mimic the requirements of the underlying interaction pattern. For example, the queue of actions can be represented by a *PropertyAffordance*. [BACnet binding](https://w3c.github.io/wot-binding-templates/bindings/protocols/bacnet/index.html#event-mappings) is currently also expressing relationships between events and actions, thanks to a combination of ad hoc affordances and binding of defined well-known relationships. 
```

- **Support for queue**: Programs can be queued and managed by the server, allowing multiple programs to be executed in sequence or based on priority.
- **Reference**: https://reference.opcfoundation.org/Core/Part10/v105/docs/4.2

#### OpenFlexure Microscope

- **Support for Asynchronous Actions**: Yes
- **Description**: The OpenFlexure Microscope provides a RESTful API that allows users to control the microscope and perform various actions. Some actions, such as capturing images or moving the stage, can be long-running and are designed to be asynchronous. When an action is initiated, the API returns a job ID that can be used to monitor the status of the action. Users can query the status of the job and retrieve results once the action is complete. There is also support for cancelling ongoing actions if needed.
- **Support for queue**: Yes, the API supports queuing of actions, allowing multiple actions to be initiated and managed independently.
- **Reference**: [TD](https://github.com/w3c/wot-testing/tree/main/events/2024.11.Munich/TDs/openflexure) and [presentation](https://www.youtube.com/watch?v=TI6HUOw6lhU)

> ### Managiable Events in IoT Platforms and Frameworks

Reviewer: ```suggestion
### Manageable Events in IoT Platforms and Frameworks
```

Sadly, it seems I can't disclose any information about Philips Hue Smart API as explained in the [terms and conditions](https://developers.meethue.com/terms-of-use-and-conditions/) of the API documentation:
> Confidentiality: You shall not disclose to any person the Developer Content or any other information of a confidential nature provided to you by Signify, including Feedback (“Confidential Information”), except that you may disclose the Confidential Information: (i) to employees, officers, representatives or advisers of the company you work for and who need to know such information for the purposes of carrying out your obligations under these Terms (“Representatives”), provided that the Representatives comply with the confidentiality obligations under this clause; and (ii) as may be required by law, court order or any governmental or regulatory authority. You shall not use the Confidential Information for any purpose other than to perform your obligations under these Terms.

### Shelly Devices

- **Support for Asynchronous Actions**: Yes, sort of.

> - **Description**: Device interaction is based on an RPC Protocol (specifically [JSON-RPC 2.0](https://www.jsonrpc.org/specification)). The device usually executes the command and immediately returns a response. While subsequent state changes can be monitored through notifications send over MQTT or Websockets there is no built-in concepts that encapsulate a long-running, monitorable process.

Reviewer 3: ```suggestion
- **Description**: Device interaction is based on an RPC Protocol (specifically [JSON-RPC 2.0](https://www.jsonrpc.org/specification)). The device usually executes the command and immediately returns a response. While subsequent state changes can be monitored through notifications sent over MQTT or Websockets, there are no built-in concepts that encapsulate a long-running, monitorable process.
```

- **Support for queue**: Yes, but it is not clear if there is a querible global queue per device. 
- **Reference**: [SmartThings Execute a command](https://developer.smartthings.com/docs/api/public#tag/Devices/operation/executeDeviceCommands) 
### Tuya IoT

- **Support for Asynchronous Actions**: Unclear
- **Description**: Tuya's interaction pattern is based on a state synchronization model. Consumers send commands to change a device's "desired" state, and the device reports its "current" state. Although it is possible to send commands or actions in batch, the documentation suggests that those are executed syncronously and the result is returned immediatly or fail if the device is not online. 
- **Support for queue**: No.
- **Reference**: [Send Commands](https://developer.tuya.com/en/docs/cloud/e2512fb901?id=Kag2yag3tiqn5) [Send Actions](https://developer.tuya.com/en/docs/cloud/687123828c?id=Kcp2kw4igv7l8)

## Summarized Problem

> As introduced, the current Thing Description interaction model does not take into account the complex interaction pattern used in well-known applications and frameworks. In particular, action affordances fail to describe real-world actions that take time and have an impact on the physical world. Moreover, some scenarios require to describe (currently unbounded) realtionships between affordances that can result in specific interactions that a Cusumer can or have to perform to correctly interact with the remote Thing. 

Reviewer 3: ```suggestion
As introduced, the current Thing Description interaction model does not take into account the complex interaction patterns used in well-known applications and frameworks. In particular, action affordances fail to describe real-world actions that take time and have an impact on the physical world. Moreover, some scenarios require description of (currently unbounded) relationships between affordances that can result in specific interactions that a Customer can or is required to perform to correctly interact with the remote Thing. 
```

Possible challenges:
- There is no definitive "state machine" for manageable actions; different solutions choose different state names or semantics
- Affordance relationships are currently unbounded. Even if we have concrete examples (BACnet and Webthings), relationships might involve:
    * Just logical reference:
        - Affordance A is a collection of instances of B
        - The payloads of Affordance A and B are somehow related (An action influences a Property)
        - ...

>     * Call order (e.g. Affordance B can be called after A completes)
>     * Call order and Payload dependency (e.g. Affordance B requires data from Affordance A)

Reviewer 3: ```suggestion
    * Call order (e.g., Affordance B can be called after A completes)
    * Call order and Payload dependency (e.g., Affordance B requires data from Affordance A)
```
